tortoisebeacon/handlers.go

```diff
@@ -22,95 +23,92 @@ const TBFirstVotingProtocol = "TBFirstVotingGossip"
 
 		return
 	}
 
-	if err := tb.handleProposalMessage(ctx, m); err != nil {
-		tb.Log.With().Error("Failed to handle proposal message",
-			log.String("sender", data.Sender().String()),
-			log.String("message", m.String()),
-			log.Err(err))
+	currentEpoch := tb.currentEpoch()
+	if message.EpochID < currentEpoch {
+		tb.Log.With().Debug("Received proposal message from previous epoch, ignoring",
+			log.Uint64("message_epoch", uint64(message.EpochID)),
+			log.Uint64("current_epoch", uint64(currentEpoch)))
 
 		return
 	}
 
-	data.ReportValidation(ctx, TBProposalProtocol)
-}
+	tb.proposalChansMu.Lock()
+	ch := tb.ensureProposalChanExists(message.EpochID)
+	tb.proposalChansMu.Unlock()
 
-func (tb *TortoiseBeacon) handleProposalMessage(ctx context.Context, m ProposalMessage) error {
-	// TODO(nkryuchkov): buffer messages that came from future epochs because of time drifts on different nodes (https://github.com/spacemeshos/go-spacemesh/pull/2529#issuecomment-877772213)
+	extendedMessage := extendedProposalMessage{
+		message: message,
+		gossip:  data,
+	}
 
-	receivedTimestamp := time.Now()
+	go func() {
+		select {
+		case <-tb.CloseChannel():
+		case ch <- extendedMessage:
+		}
+	}()
+}
+
+func (tb *TortoiseBeacon) handleProposalMessage(m ProposalMessage) error {
 	currentEpoch := tb.currentEpoch()
 
 	atxID, err := tb.atxDB.GetNodeAtxIDForEpoch(m.MinerID, currentEpoch-1)
-	if err != nil {
+	if errors.Is(err, database.ErrNotFound) {
 		tb.Log.With().Warning("Miner has no ATXs in the previous epoch")
+
 		return nil
 	}
 
-	currentEpochProposal, err := tb.buildProposal(currentEpoch)
 	if err != nil {
-		return fmt.Errorf("calculate proposal: %w", err)
+		return fmt.Errorf("get node ATXID for epoch: %w", err)
 	}
 
-	sender := m.MinerID.VRFPublicKey // hash of a miner PK
-	ok := tb.vrfVerifier(sender, currentEpochProposal, m.VRFSignature)
-	if !ok {
-		// TODO(nkryuchkov): attach telemetry
-		tb.Log.With().Warning("Received malformed proposal message: VRF is not verified",
-			log.String("sender", m.MinerID.Key))
-
-		// TODO(nkryuchkov): add a test for this case
-		return ErrMalformedProposal
+	if done, err := tb.verifyProposalMessage(m, currentEpoch); done {
+		return fmt.Errorf("verify proposal message: %w", err)
 	}
 
-	epochWeight, _, err := tb.atxDB.GetEpochWeight(currentEpoch)
-	if err != nil {
-		return fmt.Errorf("get epoch weight: %w", err)
+	if err := tb.classifyProposalMessage(m, atxID, currentEpoch); err != nil {
+		return fmt.Errorf("classify proposal message: %w", err)
 	}
 
-	passes, err := tb.proposalPassesEligibilityThreshold(m.VRFSignature, epochWeight)
-	if err != nil {
-		return fmt.Errorf("proposalPassesEligibilityThreshold: %w", err)
-	}
+	return nil
+}
 
-	if !passes {
-		tb.Log.With().Warning("Rejected proposal message which doesn't pass threshold")
-		return nil
-	}
+func (tb *TortoiseBeacon) classifyProposalMessage(m ProposalMessage, atxID types.ATXID, currentEpoch types.EpochID) error {
+	receivedTimestamp := time.Now()
```

Comment:
shouldn't we use adjusted clock?

Reply:
@dshulyak What do you mean by the adjusted clock? Actually, the timestamp should be gotten earlier, I implemented it here: https://github.com/spacemeshos/go-spacemesh/pull/2553/commits/ee07b71b510bd57fa4ff397083f6d1ca35cb4d61. But I guess you meant something different?

Reply:
my bad, i thought that we are adjusting our local clock with ntp, but instead we only check that time.Now is in sync with NTP